src/apps/carbon-defi/ethereum/carbon-defi.strategy.contract-position-fetcher.ts

```diff
@@ -0,0 +1,142 @@
+import { ETH_ADDR_ALIAS, ZERO_ADDRESS } from '~app-toolkit/constants/address';
+import { PositionTemplate } from '~app-toolkit/decorators/position-template.decorator';
+import { drillBalance } from '~app-toolkit/helpers/drill-balance.helper';
+import { ContractPositionBalance } from '~position/position-balance.interface';
+import { ContractPosition, MetaType } from '~position/position.interface';
+import { ContractPositionTemplatePositionFetcher } from '~position/template/contract-position.template.position-fetcher';
+import {
+  DefaultContractPositionDefinition,
+  UnderlyingTokenDefinition,
+  GetTokenBalancesParams,
+  GetTokenDefinitionsParams,
+  GetDisplayPropsParams,
+  GetDataPropsParams,
+} from '~position/template/contract-position.template.types';
+
+import { CarbonDefiViemContractFactory } from '../contracts';
+import { CarbonController } from '../contracts/viem/CarbonController';
+
+interface Order {
+  y: bigint;
+  z: bigint;
+  A: bigint;
+  B: bigint;
+}
+
+interface Strategy {
+  id: bigint;
+  owner: string;
+  tokens: readonly [string, string];
+  orders: readonly [Order, Order];
+}
+
+interface StrategyDefinition extends DefaultContractPositionDefinition {
+  strategy: Strategy;
+}
+
+type StrategyProps = {
+  strategy: Strategy;
+};
+
+function isActiveStrategy(strategy: Strategy) {
+  const [buy, sell] = strategy.orders;
+  return !!buy.A || !!buy.B || !!sell.A || !!sell.B;
+}
+
+@PositionTemplate()
+export class EthereumCarbonDefiStrategyContractPositionFetcher extends ContractPositionTemplatePositionFetcher<
+  CarbonController,
+  StrategyProps,
+  StrategyDefinition
+> {
+  groupId = 'strategy';
+  groupLabel = 'Carbon Defi';
+
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(CarbonDefiViemContractFactory) protected readonly factory: CarbonDefiViemContractFactory,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract() {
+    return this.factory.carbonController();
+  }
+
+  override async getDataProps(params: GetDataPropsParams<CarbonController, StrategyProps, StrategyDefinition>) {
+    return { strategy: params.definition.strategy };
+  }
+
+  override async filterPositionsForAddress(
+    address: string,
+    positions: ContractPosition<StrategyProps>[],
+  ): Promise<ContractPosition<StrategyProps>[]> {
+    return positions.filter(position => {
+      const owner = position.dataProps.strategy.owner;
+      return owner.toLowerCase() === address.toLowerCase();
+    });
+  }
+
+  // Override the getBalance query to keep token with budget set to 0
+  override async getBalances(address: string): Promise<ContractPositionBalance<StrategyProps>[]> {
+    const multicall = this.appToolkit.getViemMulticall(this.network);
+    if (address === ZERO_ADDRESS) return [];
+
+    const contractPositions = await this.getPositionsForBalances();
+    const filteredPositions = await this.filterPositionsForAddress(address, contractPositions);
```

Comment:
Performance wise, it doesn't make sense to loop through the list of pairs every time to fetch the balance for a given wallet

Reply:
I'm not sure to see the performance issue: 
- I can filter inside the map and run `compact` after, but it'll be the same amount of iteration.
- I can do a query to `tokensByOwners` to have the list, but I'll still need `getPositionsForBalances` to use `drillBalance`. So we trade a sync filter for a call to the network. So performance wise we loose.
- I can keep it the result in a property on the class, but if the user transfer his strategy it'll break.

Overall I think this is the best performance I can have. Also it's how the contract position template seems to [work currently](https://github.com/bancorprotocol/zapper-studio/blob/main/src/position/template/contract-position.template.position-fetcher.ts#L227).

Reply:
I mentioned a performance issue mainly because 
`ContractPositionTemplatePositionFetcher` is extended over `CustomAppContractPositionFetcher` which at worst will result in N^2 (n = number of definitions)  in term of complexity. 

ContractPositionTemplatePosition is basic and works for 95% of positions/balances (see typical example `EthereumApi3StakingContractPositionFetcher`)

Whenever you can't get the user's balances for specific position, you will have to use `CustomAppContractPositionFetcher` (e.g BaseAerodromeBribeContractPositionFetcher)

I'll admit "performance issue" was a bit too intense of a wording.. my bad.

Reply:
What I would do is
- Extend `CustomPositionTemplatePositionFetcher` (have a look at `BaseAerodromeBribeContractPositionFetcher`, especially at the code within `VotingRewardsContractPositionFetcher`
```
  getTokenBalancesPerPosition(): never {
    throw new NotImplementedException();
  }
```
Then, since it's a custom cp, you will have to implement
`getRawBalances`